### PR TITLE
Support aggregated endpoint for customized stoppable check

### PR DESCRIPTION
**Pull request by MarkGaox**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
#2918

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

* This PR supports the aggregated customized stoppable check.
* When configuring the RESTConfig with an specific URI (e.g., "https://*:123"), users means to use the aggregated health check instead of the instance/partition level checks. 
* The aggregated health check will pass `instances`, `toBeStoppedInstances`, and the `clusterID` to the health check endpoint and expect the user to return the response regarding whether those instances are stoppable and the reason of being not stoppable. 
* Support Redirect of request if receiving status code of 302 with max retry of 3 times.
* Following is an expected request + response:
```
     example url: POST http://<baseUrl>/aggregatedHealthStatus -d {
       "instances" : ["n1", "n3", "n5", "n9"],
       "to_be_stopped_instances": "["n2", "n4"]",
       "cluster_id": "cluster1",
       "<key>": "<value>"
     ...
     }
     example response:
     {
       "stoppableInstances": ["n1", "n3"],
       "nonStoppableInstancesWithReasons": {
       "n5": "reason1,reason2",
       "n9": "reason3"
     }
```

### Tests

- [X] The following tests are written for this issue:

 ```
mvn test -Dtest=TestInstancesAccessor,TestMaintenanceManagementService,TestInstanceValidationUtilInRest,TestPerInstanceAccessor,TestCustomRestClient -pl helix-rest && mvn test -Dtest=TestInstanceValidationUtil -pl helix-core
```

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Comment by xyuanlu** [comment]
could you please also update naming in pr?

File: helix-core/src/main/java/org/apache/helix/model/RESTConfig.java [hunk 1]
```diff
@@ -83,4 +85,18 @@ public String getBaseUrl(String instance) {
...
     return baseUrl.replace("*", instanceVip);
   }
+
+  /**
+   * Retrieves the configured health URL if no wildcard is present.
+   * Returns an empty Optional if the URL contains a wildcard.
+   *
+   * @return Optional containing the exact configured URL, or empty if a wildcard is present.
+   */
+  public Optional<String> getConfiguredHealthUrl() {
```
Comment (zpinto) [inline helix-core/src/main/java/org/apache/helix/model/RESTConfig.java:95]: Can we think of a more descriptive name for this?

Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/model/RESTConfig.java:95]: `getCompleteConfiguredHealthUrl`

File: helix-core/src/main/java/org/apache/helix/model/RESTConfig.java [hunk 2]
```diff
@@ -83,4 +85,18 @@ public String getBaseUrl(String instance) {
 
     return baseUrl.replace("*", instanceVip);
   }
+
+  /**
+   * Retrieves the configured health URL if no wildcard is present.
```
Comment (xyuanlu) [inline helix-core/src/main/java/org/apache/helix/model/RESTConfig.java:90]: I would suggest include more detail about aggregated healthcheck url and individual health check url. What are the differences, example etc.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java [hunk 3]
```diff
@@ -110,42 +152,102 @@ protected JsonNode getJsonObject(HttpResponse httpResponse) throws IOException {
...
       }
+    }
 
-      return response;
-    } catch (IOException e) {
-      // Release connection to be reused and avoid connection leakage.
-      postRequest.releaseConnection();
-      throw e;
+    if (retries >= MAX_REDIRECT) {
+      throw new IOException("Maximum redirect retries reached.");
```
Comment (xyuanlu) [inline helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:217]: Why throw exp here?

Comment (MarkGaox) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:217]: Ya, it's probably not needed. We can just return the response in that case and the method `handleResponse` will generate the proper exception.
Will remove it.

File: helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java [hunk 4]
```diff
@@ -93,15 +98,52 @@ public Map<String, Boolean> getPartitionStoppableCheck(String baseUrl, List<Stri
...
-          kv -> result.put(kv.getKey(), kv.getValue().get(IS_HEALTHY_FIELD).asBoolean()));
-      return result;
-    };
+    JsonConverter<Boolean> jsonConverter = this::extractPartitionBooleanMap;
     return handleResponse(post(url, payLoads), jsonConverter);
   }
 
+  @Override
+  public Map<String, List<String>> getAggregatedStoppableCheck(String baseUrl,
+      List<String> instances, Set<String> toBeStoppedInstances, String clusterId,
```
[the remark this answers is not in the document]
Comment (MarkGaox) [reply helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java:107]: Thanks for catching it. It should use clusterId all the time.

**Review by zpinto — APPROVED** [review]
LGTM! Thanks for the change :)

**Review by junkaixue — COMMENTED** [review]
I don't have major concerns. But take a look of failure tests.

**Comment by MarkGaox** [comment]
> I don't have major concerns. But take a look of failure tests.
The failed test passed locally. 

Screenshot 2024-09-25 at 1 30 42 AM

**Comment by MarkGaox** [comment]
This PR is approved by @zpinto and @junkaixue. TFTR. It's ready to be merged. The commit message: "Support aggregated endpoint for customized stoppable check"

**Pull request merged**